sdc-plugin/clocks.cc

```diff
@@ -0,0 +1,255 @@
+    auto clock_wires = clock.GetClockWires();
+    for (auto clock_wire : clock_wires) {
+#ifdef SDC_DEBUG
+	log("Clock wire %s\n", RTLIL::unescape_id(clock_wire->name).c_str());
+#endif
+	auto buf_wires = pass->FindSinkWiresForCellType(clock_wire, buffer.name,
+	                                                buffer.output);
+	int path_delay(0);
+	for (auto wire : buf_wires) {
+#ifdef SDC_DEBUG
+	    log("%s wire: %s\n", buffer.name.c_str(),
+	        RTLIL::unescape_id(wire->name).c_str());
+#endif
+	    path_delay += buffer.delay;
+	    AddClock(RTLIL::unescape_id(wire->name), wire, clock.Period(),
+	             clock.RisingEdge() + path_delay,
+	             clock.FallingEdge() + path_delay);
+	}
+    }
+}
+
+void Clocks::WriteSdc(std::ostream& file) {
+    for (auto& clock : clocks_) {
+	auto clock_wires = clock.GetClockWires();
+	// FIXME: Input port nets are not found in VPR
+	if (std::all_of(clock_wires.begin(), clock_wires.end(),
+	                [&](RTLIL::Wire* wire) { return wire->port_input; })) {
+	    continue;
+	}
+	file << "create_clock -period " << clock.Period();
+	file << " -waveform {" << clock.RisingEdge() << " "
+	     << clock.FallingEdge() << "}";
+	for (auto clock_wire : clock_wires) {
+	    if (clock_wire->port_input) {
+		continue;
+	    }
+	    file << " " << Clock::ClockWireName(clock_wire);
+	}
+	file << std::endl;
+    }
+}
+
+Clock::Clock(const std::string& name, RTLIL::Wire* wire, float period,
+             float rising_edge, float falling_edge)
+    : name_(name),
+      period_(period),
+      rising_edge_(rising_edge),
+      falling_edge_(falling_edge) {
+    AddWire(wire);
+}
+
+Clock::Clock(const std::string& name, std::vector<RTLIL::Wire*> wires,
+             float period, float rising_edge, float falling_edge)
+    : name_(name),
+      period_(period),
+      rising_edge_(rising_edge),
+      falling_edge_(falling_edge) {
+    std::for_each(wires.begin(), wires.end(),
+                  [&, this](RTLIL::Wire* wire) { AddWire(wire); });
+}
+
+void Clock::AddWire(RTLIL::Wire* wire) {
+    auto clock_wire = std::find(clock_wires_.begin(), clock_wires_.end(), wire);
+    if (clock_wire == clock_wires_.end()) {
+	clock_wires_.push_back(wire);
+    }
+}
+
+void Clock::UpdateClock(RTLIL::Wire* wire, float period, float rising_edge,
+                        float falling_edge) {
+    UpdateWires(wire);
+    UpdatePeriod(period);
+    UpdateWaveform(rising_edge, falling_edge);
+}
+
+void Clock::UpdateWires(RTLIL::Wire* wire) {
+    if (std::find(clock_wires_.begin(), clock_wires_.end(), wire) ==
+        clock_wires_.end()) {
+	clock_wires_.push_back(wire);
+    }
+}
+
+void Clock::UpdatePeriod(float period) {
+    period_ = period;
+    rising_edge_ = 0;
+    falling_edge_ = period / 2;
+}
+
+void Clock::UpdateWaveform(float rising_edge, float falling_edge) {
+    rising_edge_ = rising_edge;
+    falling_edge_ = falling_edge;
+    assert(falling_edge - rising_edge == period_ / 2);
```

Comment:
This probably needs to be a delta with error, like `((falling_edge - rising_edge) - (period_ / 2)) <= period_*(10*std::numeric_limits<decltype(period_)>.epsilon())`

Reply:
This should probably not be an `assert`, as this can be compiled out.  Probably just a `if violated, raise error`?

Reply:
In my opinion we don't a check at all. I mean after all this plugin is a SDC commands "forwarder" which takes the users SDC and creates another one based on this input. The real consumer is VPR and I believe there are checks there to parse and validate these values. I don't think the plugin is the place the perform such checks. Besided non 50% duty cycle is supported in VPR, however according to the documentation it's being treated as 50% anyways.

Reply:
So something to watch out for is if a non-0 deg phase is put through a buffer or PLL, then output clock signal should respect the phase.

To be more concrete:
 - Buffers (like IBUF and BUFG) should output a clock with the same frequency and phase shift equal to the delay through the buffer
 - Clock synth's (like MMCM and PLL) should output a clock with the computed frequency, and a phase shift equal to the phase specified.  For example if the PLL input has a 20 degree phase shift, and the request to the PLL is to add another 45 degree phase shift, the output waveform should have a 65 degree phase shift.

Does that make sense?

Reply:
It does make sense. The buffers are already implemented this way, for PLLs we currently have the delay set to 0, so on the output we want the phase of the input + the requested shift + delay (equal to 0). Is this correct?

Reply:
> It does make sense. The buffers are already implemented this way, for PLLs we currently have the delay set to 0, so on the output we want the phase of the input + the requested shift + delay (equal to 0). Is this correct?

I believe so.  Something to watch for is what happens when the total phase shift is > 180 degrees.